### PR TITLE
change domain thread name setting to be more portable and best-effort

### LINKED ISSUE #10870: caml_thread_setname compilation and adjustments
I am opening this issue to track two things that were brought up during the Multicore PR review. (#10831)

The first item is a comment made by @damiendoligez about `caml_ml_domain_set_name` not taking into account the extra id appended by `caml_thread_setname` when checking for length. (see https://github.com/ocaml/ocaml/pull/10831#discussion_r779668302)

I think `caml_ml_domain_set_name` should actually not append an id to the thread name: this function is the user facing utility to allow user to name the underlying domain's system thread, and as such users should have full control over the name. This also makes the length checking logic simpler.

The second item is an observation made by @avsm on the macro used to check for the existence of `pthread_setname_np` and portability concerns.
I could not track down the exact comment, but I think it is worth discussing this here as well.
It seems like @avsm was unsure about the usage of `_GNU_SOURCE` for the feature check and that we may have to be a bit more specific for it to be not bricky.

### PR DESCRIPTION
**Pull request by avsm**
- change caml_set_domain_name to best-effort and a void return
- shorten backup thread name to just "Backup" to avoid hitting the
  15-character limit when domain ids are appended
- add cases for *BSD permutations of the functions

There is an unfortunate portability mess around setting pthread names, so we now use the various permutations on FreeBSD and OpenBSD (which require another header include as well) and NetBSD (which is similar to Linux, but with a different error code set).

The code previously used GNU_SOURCE to detect Linux, which isnt much use as GNU_SOURCE is unconditionally defined at the top of unix.c, so we just assume that the 'default' platform is Linux. pthread_setname_np is only defined in glibc 2.12 or higher, but looking for that specifically makes it complex to also detect musl, so it's simplest to leave it as the default case.

Since setting the thread name is always best-effort and used for debugging, there are no new configure tests added.

Fixes #10870 and FreeBSD/amd64 passes the test suite with this change. OpenBSD now builds but requires more runtime fixes (to follow).

### CODE REVIEW

**Review by mndrix — COMMENTED** [review]
The commit message mentions `caml_set_domain_name` but I think the correct name is `caml_domain_set_name`.  The rest looks good to me, FWIW.  I'll compile and test as part of #10875.

**Comment by avsm** [comment]
@engil also suggests not appending the domain id, since this should be under user control. I'll make that change too later.

**Comment by abbysmal** [comment]
@avsm Just to double check, my idea was to not have it appended when it is called through `caml_ml_domain_set_name` (the user facing function).
I think that it is useful for the "default" names to have the id pre-prended. But this involves a bit of code motion, so maybe it is out of scope for this PR?

**Comment by avsm** [comment]
@engil thanks for clarifying. The current code doesn't append the domain id from the user-facing function. The appending function is:

```
void caml_domain_set_name(char *name)
{
  char thread_name[MAX_DOMAIN_NAME_LENGTH];
  snprintf(thread_name, MAX_DOMAIN_NAME_LENGTH,
           "%s%d", name, Caml_state->id);
  caml_thread_setname(thread_name);
}
```

...which is called by the runtime to set the 'Backup%d' and 'Domain%d' thread names. Then there is:

```
CAMLprim value caml_ml_domain_set_name(value name)
{
  CAMLparam1(name);

  if (caml_string_length(name) >= MAX_DOMAIN_NAME_LENGTH)
    caml_invalid_argument("caml_ml_domain_set_name");
  caml_thread_setname(String_val(name));
  CAMLreturn(Val_unit);
}
```

which sets the name directly via `caml_thread_setname` as well. 

So I think this is ready for review/approval now. I've fixed the typo in the commit message that @mndrix identified above.

**Comment by abbysmal** [comment]
Thank you for the clarification on `caml_ml_domain_set_name`, indeed I forgot it does not append anything!

The changes looks good to me as well, good call on shortening `BackupThread` (which should buy us some time until we can boot over a million domains :-) ).

**Comment by avsm** [comment]
Thanks @engil! Could I beg some approvals from core developers to merge and unblock FreeBSD/OCaml trunk development please?

**Comment by hannesm** [comment]
I tested this PR, and on my FreeBSD 13.0 amd64 laptop, it compiles fine (and the code looks correct). The output of a testrun is:
```
Summary:
  3085 tests passed
   70 tests skipped
    0 tests failed
  179 tests not started (parent test skipped or failed)
    0 unexpected errors
  3334 tests considered
```

Which looks pretty good, thanks.

**Comment by fabbing** [comment]
I also took a look at the diff and it looks good to me!

**Review by kayceesrk — APPROVED** [review]
The changes look good to me.

**Review by gasche — APPROVED** [review]
Approved on KC's behalf.

**Pull request merged**
